### PR TITLE
refuse application

### PR DESCRIPTION
**Pull request by ofux**
- ✨ refuse application
- ✨ add refuse application endpoint
- 🔀 fix build after rebase
- ♻️ only accept one application on assign
- ⚰️ remove dead code following the refactoring

In this PR, I propose to do the following changes:
- when a contribution is assigned/claimed, the corresponding application is automatically accepted, but the others remain untouched (ie. pending)
- in the contribution aggregate, when a contribution is assigned/claimed, the corresponding applicant is removed from the list of applicants. In other words, the list of applicants only contains _pending_ applications. This makes further checks easier: when someone apply, you just need to check he is not already in the list of applicants, and when someone refuse an application, you just need to check the applicant is in the list.

### CODE REVIEW

File: marketplace-domain/src/contribution/aggregate_root/mod.rs [hunk 1]
```diff
@@ -90,6 +106,18 @@ impl EventSourcable for Contribution {
 					applicants.push(contributor_id.clone());
 					Self { applicants, ..self }
 				},
+				ContributionEvent::ApplicationRefused {
+					id: _,
+					contributor_id,
+				} => {
+					let mut applicants = self.applicants;
+					if let Some(index) =
```
Comment (AnthonyBuisset) [inline marketplace-domain/src/contribution/aggregate_root/mod.rs]: I would have used iter.filter().collect(), but this works as well

File: marketplace-domain/src/contribution/projectors/application.rs [hunk 2]
```diff
@@ -216,8 +233,91 @@ mod tests {
...
+		application_projection_repository.expect_update().never();
+		application_projection_repository
+			.expect_create()
+			.with(eq(ApplicationProjection::new(
+				application_id,
+				contribution_id.clone(),
+				contributor_id.clone(),
+			)
+			.into_refused()))
+			.once()
```
Comment (AnthonyBuisset) [inline marketplace-domain/src/contribution/projectors/application.rs:305]: nice one !

File: marketplace-core/src/application/contribution/refuse_application.rs [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+					}
+				} else {
+					panic!("Contribution event expected");
+				}
+			})
+			.collect();
+		self.event_store.append(&contribution_id, storable_events)?;
+		// TODO: the usecase shouldn't know about the projectors, it should just push the events to
+		// a bus
+		for event in &events {
```
Comment (AnthonyBuisset) [inline marketplace-core/src/application/contribution/refuse_application.rs:88]: we definitely need an event bus !

File: marketplace-core/src/application/contribution/refuse_application.rs [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+			self.application_projector.on_event(event).await;
+			self.contributor_projector.on_event(event).await;
```
Comment (AnthonyBuisset) [inline marketplace-core/src/application/contribution/refuse_application.rs:90]: I do not think this one is needed

**Review by AnthonyBuisset — APPROVED** [review]

File: marketplace-domain/src/contribution/aggregate_root/mod.rs [hunk 5]
```diff
@@ -101,23 +117,11 @@ impl EventSourcable for Contribution {
 				ContributionEvent::Applied {
 					id: _,
 					contributor_id,
-				} => {
-					let mut applicants = self.applicants;
-					applicants.push(contributor_id.clone());
-					Self { applicants, ..self }
-				},
+				} => self.add_applicant(contributor_id),
```
Comment (AnthonyBuisset) [inline marketplace-domain/src/contribution/aggregate_root/mod.rs:120]: I would rename those functions to `with_applicant` and `without_applicant` as you return a modified `Self`

File: marketplace-domain/src/contribution/aggregate_root/mod.rs (hunk 1 above)
Comment (ofux) [reply marketplace-domain/src/contribution/aggregate_root/mod.rs]: I really wanted to find the last index (there should always be only one, but we shouldn't rely on this here), and found it easier to write it this way.

File: marketplace-core/src/application/contribution/refuse_application.rs (hunk 3 above)
Comment (ofux) [reply marketplace-core/src/application/contribution/refuse_application.rs:88]: Yes! Yes! Yes!

File: marketplace-domain/src/contribution/aggregate_root/mod.rs (hunk 5 above)
Comment (ofux) [reply marketplace-domain/src/contribution/aggregate_root/mod.rs:120]: agreed

File: marketplace-core/src/application/contribution/refuse_application.rs (hunk 4 above)
Comment (ofux) [reply marketplace-core/src/application/contribution/refuse_application.rs:90]: indeed

File: marketplace-domain/src/contribution/projectors/contribution/contribution_with_github_data_projector.rs [hunk 6]
```diff
@@ -109,6 +109,7 @@ impl EventListener for WithGithubDataProjector {
 				ContributionEvent::Unassigned { id } => self.on_unassign(id),
 				ContributionEvent::Validated { id } => self.on_validate(id),
 				ContributionEvent::Applied { .. } => return,
+				ContributionEvent::ApplicationRefused { .. } => return,
```
Comment (Bernardstanislas) [inline marketplace-domain/src/contribution/projectors/contribution/contribution_with_github_data_projector.rs:112]: maybe we can group all the "bypassed" events with a `|`?

File: marketplace-domain/src/contribution/aggregate_root/mod.rs [hunk 7]
```diff
@@ -75,13 +75,13 @@ impl Contribution {
 		&self.status
 	}
 
-	fn add_applicant(self, contributor_id: &ContributorId) -> Self {
+	fn with_applicant(self, contributor_id: &ContributorId) -> Self {
```
Comment (Bernardstanislas) [inline marketplace-domain/src/contribution/aggregate_root/mod.rs:78]: Very clear!!

**Review by Bernardstanislas — APPROVED** [review]

File: marketplace-domain/src/contribution/projectors/contribution/contribution_with_github_data_projector.rs (hunk 6 above)
Comment (ofux) [reply marketplace-domain/src/contribution/projectors/contribution/contribution_with_github_data_projector.rs:112]: good idea!

**Pull request merged**